compiler/src/dotty/tools/dotc/core/TypeUtils.scala

```diff
@@ -127,8 +127,17 @@ class TypeUtils:
         case Some(types) => TypeOps.nestedPairs(types)
         case None => throw new AssertionError("not a tuple")
 
-    def namedTupleElementTypesUpTo(bound: Int, normalize: Boolean = true)(using Context): List[(TermName, Type)] =
+    def namedTupleElementTypesUpTo(bound: Int, derived: Boolean, normalize: Boolean = true)(using Context): List[(TermName, Type)] =
       (if normalize then self.normalized else self).dealias match
+        // for desugaring and printer, ignore derived types to avoid infinite recursion in NamedTuple.unapply
```

Comment:
Can we avoid the duplication with the extractor? Maybe have two extractors `NamedTuple.unapply` and `NamedTupleDirect.unapply` that share the same logic?